Sources/Internal/WebSocketConnection.swift

```diff
@@ -4,28 +4,42 @@
 
 import Foundation
 import Starscream
+import Network
+
+#if os(iOS)
+import UIKit
+#endif
+
 
 class WebSocketConnection {
     let url: WCURL
     private let socket: WebSocket
+    
+    private var isConnected: Bool = false
+    
     private let onConnect: (() -> Void)?
     private let onDisconnect: ((Error?) -> Void)?
     private let onTextReceive: ((String) -> Void)?
     // needed to keep connection alive
     private var pingTimer: Timer?
     // TODO: make injectable on server creation
     private let pingInterval: TimeInterval = 30
-
+    
     private var requestSerializer: RequestSerializer = JSONRPCSerializer()
     private var responseSerializer: ResponseSerializer = JSONRPCSerializer()
-
+    
     // serial queue for receiving the calls.
     private let serialCallbackQueue: DispatchQueue
-
+    
+#if os(iOS)
+    private var backgroundNotificationObserver: Any?
+    private var foregroundNotificationObserver: Any?
+#endif
```

Comment:
I think we don't need to retain observers since the observation will exist for the duration of this class

Reply:
Technically you're right, as blocks do not capture `self`, still, it will let our blocks passed to NotificationCenter be alive for an undefined amount of time. And contradicts the RAII principle, and leaves hanging memory in heap.

Reply:
https://developer.apple.com/documentation/foundation/notificationcenter/1411723-addobserver
`block
The block that executes when receiving a notification.

The notification center copies the block. The notification center strongly holds the copied block until you remove the observer registration.

The block takes one argument: the notification.`